test/hotspot/jtreg/runtime/Monitor/MonitorUsedDeflationThresholdTest.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStreamReader;
+import java.util.regex.Pattern;
+*/
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.Platform;
+import jdk.test.lib.process.ProcessTools;
+
+/*
+ * @test
+ * @bug 8226416
+ * @summary Test the MonitorUsedDeflationThreshold and NoAsyncDeflationProgressMax options.
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib
+ * @run driver MonitorUsedDeflationThresholdTest
+ */
+
+public class MonitorUsedDeflationThresholdTest {
+    public static final int DELAY_SECS = 10;
+    public static int inflate_count = 0;
+
+    public static void do_work(int count) {
+        System.out.println("Recursion count=" + count);
+        if (count > inflate_count) {
+            System.out.println("Exceeded inflate_count=" + inflate_count);
+
+            System.out.println("Delaying for " + DELAY_SECS + " secs.");
+            try {
+                Thread.sleep(DELAY_SECS * 1000);
+            } catch (InterruptedException ie) {
+                // ignore InterruptedException
+            }
+            System.out.println("Done delaying for " + DELAY_SECS + " secs.");
+            return;
+        }
+
+        Object obj = new Object();
```

Comment:
This is a thread-local object, which means the synchronization has no affect, so I think C2 could elide the following sync-block. Using a static field would make it less likely that this can happen.

Reply:
Fixed.

Reply:
But it has to be a static array of Objects since we want a new
ObjectMonitor per recursion.